Packages/com.unity.inputsystem/InputSystem/Devices/NoiseFilter.cs

```diff
@@ -0,0 +1,246 @@
+            /// </summary>
+            /// <param name="noiseFilterBuffer">The Noisefilter buffer for doing whole control filtering.</param>
+            /// <param name="device">The device you want to apply filtering to.</param>
+            public void Apply(IntPtr noiseFilterBuffer, InputDevice device)
+            {
+                if (controlIndex >= device.allControls.Count)
+                    throw new IndexOutOfRangeException("NoiseFilter has array index beyond total size of device's controls");
+
+                InputControl control = device.allControls[controlIndex];
+                BitmaskHelpers.Blacklist(noiseFilterBuffer, control);
+            }
+
+            /// <summary>
+            /// Checking an individual input event for significant InputControl changes.
+            /// </summary>
+            /// <param name="inputEvent">The input event being checked for changes</param>
+            /// <param name="device">The input device being checked against </param>
+            /// <returns>True if any changes exist in the event once the device has been filtered through for noise and non-significant changes.  False otherwise.</returns>
+            public bool HasValidData(InputEventPtr inputEvent, InputDevice device)
+            {
+                if (!inputEvent.valid)
+                    throw new ArgumentException("Invalid or unset event being checked.", "inputEvent");
+
+                if (device == null)
+                    throw new ArgumentException("No device passed in to check if inputEvent has valid data", "device");
+
+                InputControl control = device.allControls[controlIndex];
+                if (control != null)
+                {
+                    switch (type)
+                    {
+                        case EElementType.TypeFloat:
+                        {
+                            InputControl<float> castedControl = control as InputControl<float>;
+                            if (castedControl != null)
+                            {
+                                float value;
+                                if (castedControl.ReadValueFrom(inputEvent, out value) && Mathf.Abs(value) > float.Epsilon)
+                                    return true;
+                            }
+                        }
+                        break;
+                        case EElementType.TypeVec2:
+                        {
+                            InputControl<Vector2> castedControl = control as InputControl<Vector2>;
+                            if (castedControl != null)
+                            {
+                                Vector2 value;
+                                if (castedControl.ReadValueFrom(inputEvent, out value) && Vector2.SqrMagnitude(value) > float.Epsilon)
+                                    return true;
+                            }
+                        }
+                        break;
```

Comment:
This logic seems like it should be *on* the control rather than having type switches here. Maybe `InputControl.IsAtDefaultValue()` or something. Unfortunately, the virtual calls are starting to accumulate in there but oh well.